store/tikv/latch/latch.go

```diff
@@ -0,0 +1,198 @@
+type Latch struct {
+	hasWaiting bool
+	// The startTS of the transaction which is the head of waiting transactions.
+	head        uint64
+	maxCommitTS uint64
+	sync.Mutex
+}
+
+// acquire tries to get current key's lock for the transaction with startTS.
+// success is true when success
+// timeout is true when the startTS is already timeout
+// newWait is true when current transaction is new for the current latch
+func (l *Latch) acquire(startTS uint64) (success, timeout, newWait bool) {
+	l.Lock()
+	defer l.Unlock()
+
+	if timeout = startTS <= l.maxCommitTS; timeout {
+		return
+	}
+
+	if l.hasWaiting == false {
+		l.head = startTS
+		l.hasWaiting = true
+		newWait = true
+	}
+	success = l.head == startTS
+	return
+}
+
+// release releases the transaction with startTS and commitTS from current latch,
+// and set the next transaction to head if hasNext is true.
+func (l *Latch) release(startTS, commitTS uint64, hasNext bool, nextStartTS uint64) {
+	l.Lock()
+	defer l.Unlock()
+	if startTS != l.head {
+		panic(fmt.Sprintf("invalid front ts %d, latch:%#v", startTS, l))
+	}
+	if commitTS > l.maxCommitTS {
+		l.maxCommitTS = commitTS
+	}
+	l.hasWaiting = hasNext
+	l.head = nextStartTS
+}
+
+// Lock is the locks' information required for a transaction.
+type Lock struct {
+	// The slot IDs of the latches(keys) that a startTS must acquire before being able to processed.
+	requiredSlots []int
+	// The number of latches that the transaction has acquired.
+	acquiredCount int
+	// The number of latches whose waiting queue contains current transaction.
+	waitedCount int
+	// Current transaction's startTS.
+	startTS uint64
+}
+
+// NewLock creates a new lock.
+func NewLock(startTS uint64, requiredSlots []int) Lock {
+	return Lock{
+		requiredSlots: requiredSlots,
+		acquiredCount: 0,
+		waitedCount:   0,
+		startTS:       startTS,
+	}
+}
+
+// Latches which are used for concurrency control.
+// Each latch is indexed by a slot's ID, hence the term latch and slot are used in interchangeable,
+// but conceptually a latch is a queue, and a slot is an index to the queue
+type Latches struct {
+	slots []Latch
+	// The waiting queue for each slot(slotID => slice of startTS).
+	waitingQueue map[int][]uint64
+	sync.RWMutex
+}
+
+// NewLatches create a Latches with fixed length,
+// the size will be rounded up to the power of 2.
+func NewLatches(size int) *Latches {
+	powerOfTwoSize := 1 << uint32(bits.Len32(uint32(size-1)))
+	slots := make([]Latch, powerOfTwoSize)
+	return &Latches{
+		slots:        slots,
+		waitingQueue: make(map[int][]uint64),
+	}
+}
+
+// GenLock generates Lock for the transaction with startTS and keys.
+func (latches *Latches) GenLock(startTS uint64, keys [][]byte) Lock {
+	slots := make([]int, 0, len(keys))
+	for _, key := range keys {
+		slots = append(slots, latches.hash(key))
+	}
+	sort.Ints(slots)
+	size := 0
+	for _, v := range slots {
+		if size == 0 || slots[size-1] != v {
+			slots[size] = v
+			size++
+		}
+	}
```

Comment:
This maybe easier to read.
```
	dedup := slots[:1]
	for i := 1; i < len(slots); i++ {
		if slots[i] != slots[i-1] {
			dedup = append(dedup, slots[i])
		}
	}
```

Reply:
@tiancaiamao What do you think?

Reply:
Well, @coocood is right! it's easier to read, you can change this way.

```
if len(slots) == 0 {
   return NewLock(startTS, nil)
}
```

As for me, I'd always like the C-style, for example:

* don't let append do the implicit work
* manual malloc when resize a slice
* prefer using `i := 0; i < len(); i++` style instead of `range`

It seems clumsy but less likely to went wrong.